### PR TITLE
Add `NonConstructor` to allow customization of ObjectWrap behavior

### PR DESCRIPTION
**Pull request by p120ph37**
Addresses #1123 by moving the error-throw into an overridable static method.  Default behavior remains exactly the same, but now a developer can override `NonConstructor` in their `ObjectWrap` subclass in order to define alternate behavior.  As a trivial example of behavior a developer might implement, this code in an `ObjectWrap` subclass would support the old (ugly) idiom of instantiating a class even when the function is called without the `new` keyword:
```
static Napi::Value MyObject::NonConstructor(const Napi::CallbackInfo& info) {
  std::vector<napi_value> args;
  for(size_t i = 0; i < info.Length(); i++) args.push_back(info[i]);
  return GetConstructor(info.Env()).New(args);
}
```

Which then allows this usage:
```
const MyObject = binding.MyObject;
const newConstructed = new MyObject();
const factoryConstructed = MyObject();
assert(newConstructed instanceof MyObject);
assert(factoryConstructed instanceof MyObject);
```

I've also included a new test case, which amounts to basically the above, and also tests that exceptions thrown inside the non-constructor are handled appropriately.

### CODE REVIEW

File: test/objectwrap_nonconstructor.cc [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      NAPI_THROW(Napi::Error::New(info.Env(), "an exception"), Napi::Value());
+    }
+    // Otherwise, act as a factory.
+    std::vector<napi_value> args;
+    for(size_t i = 0; i < info.Length(); i++) args.push_back(info[i]);
+    return MaybeUnwrap(GetConstructor(info.Env()).New(args));
+  }
+
+  static std::unordered_map<napi_env, Napi::FunctionReference*> constructors;
+
```
Comment (p120ph37) [inline test/objectwrap_nonconstructor.cc:23]: I know this is leaky without an env-shutdown-hook, but it's good enough to run the tests without clobbering `Napi::Addon<T>` via `env.SetInstanceData()` as would be the typical idiom.

Comment (p120ph37) [reply test/objectwrap_nonconstructor.cc:22]: Should probably add a comment to the source so nobody thinks it’s a good idea to copy paste this into a real project…

**Comment by mhdawson** [comment]
We discussed in the node-api team meeting today. 

We agreed giving developers the option makes sense.  We would like the PR to include updating the documentation as well.

We would prefer that instead of using  NonConstructor that we use CreateViaFunctionCall

**Comment by p120ph37** [comment]
Sounds reasonable to me - I'll make the requested adjustments.

**Comment by p120ph37** [comment]
I'm happy to name it `CreateViaFunctionCall` if that's the consensus, but I do feel like that's a little bit misleading because the actual action may be something other than creating an object -- that's just one possible implementation for the method.  All I intended to convey with my original `NonConstructor` name was that it handles the case where the `new` keyword was not used.  Absent the `new` keyword, the function is technically allowed to do anything it wants, including returning an object of some completely different type, or returning a "builder" object, or an object-factory, etc.  So calling it `Create...` seems a little presumptive ;-)

**Comment by vmoroz** [comment]
> I'm happy to name it `CreateViaFunctionCall` if that's the consensus, but I do feel like that's a little bit misleading because the actual action may be something other than creating an object -- that's just one possible implementation for the method. All I intended to convey with my original `NonConstructor` name was that it handles the case where the `new` keyword was not used. Absent the `new` keyword, the function is technically allowed to do anything it wants, including returning an object of some completely different type, or returning a "builder" object, or an object-factory, etc. So calling it `Create...` seems a little presumptive ;-)

+@mhdawson and @gabrielschulhof

@p120ph37, it is a great point! There were a few points for the alternative name:
- `NonConstructor` sounds a little bit alien and does not communicate the function purpose clearly.
- What if we use the same terms as in the JS docs? E.g. in [MDN Date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) they use the phrase `When called as a function, returns a string representation of the current date and time ...`. Thus, we thought that the name should have something from the `called as function` phrase.
- The method name is better start with a verb.

Since `Create...` seems to be wrong here (even the `Date()` call does not create anything), could we use something like `HandleWhenCalledAsFunction`, `OnCalledAsFunction`, or something else along these lines? E.g. to communicate that this method handles the situation when the JS function representing the native class was called as a function rather than as a constructor.

**Comment by p120ph37** [comment]
I like `OnCalledAsFunction` - it's not too verbose, and the `On` prefix is already familiar and implies the mechanism of action: the developer should define the method in order to change the default behavior.

**Comment by vmoroz** [comment]
> I like `OnCalledAsFunction` - it's not too verbose, and the `On` prefix is already familiar and implies the mechanism of action: the developer should define the method in order to change the default behavior. I'll go ahead with that name first, and we can see what the doc looks like.

I like it too! Though, the final decision is from @mhdawson and @gabrielschulhof.

File: test/objectwrap_nonconstructor.cc (hunk 1 above)
Comment (p120ph37) [reply test/objectwrap_nonconstructor.cc]: I've added an env cleanup-hook so this test-code shouldn't leak anymore and should be adaptable to real-world projects.

**Comment by mhdawson** [comment]
adding my +1 for OnCalledAsFunction

**Review by mhdawson — APPROVED** [review]
LGTM

**Comment by mhdawson** [comment]
Landed as cee899ade542d8d19429d2c32ac41f0cc5328d3a

**Pull request closed without merging**